### ISSUE TITLE
Subtitles not displaying

### ISSUE THREAD

**Issue by radinamatic** [learningequality/ka-lite#3831]
I'm having problems figuring out why no subs are showing, even though I have them downloaded, and can see them in the kalite\static\srt folder. 

**Branch**: today's develop

**Expected behavior**: Videos that have subs should show the subtitle or caption (not really sure which one, probably caption since that's the type of track in the player's code) button in the player control bar.
 
**Actual behavior**: No sub/caption button is visible in the player control bar:

![sub-button-display-none](https://cloud.githubusercontent.com/assets/1457929/8140693/46cdcaac-115e-11e5-823a-6894ae7b8346.png)

Since both buttons have `< style="display: none;" >` , I'm gonna venture a guess that the subtitle URLs for the  `<track >`  are not generated properly in the kalite/topic_tools/_ _init_ _.py file.

**Steps to reproduce**:

1. Set the DEBUG=False in the local settings
2. Choose the video that has subs from the sidebar (I checked the [APRIL 2015 MASTER](https://docs.google.com/spreadsheets/d/1k5xh2UXV3EchRHnYzeP6-YGKrmba22vsltGSuU9bL88/edit#gid=0) file, maybe there's another version?)

I tested with this file which has both English transcript & Spanish subtitle:
**Comparing numbers through 10**
comparing-numbers-through-10-example
https://www.khanacademy.org/math/early-math/cc-early-math-counting-topic/cc-early-math-comparing-numbers/v/comparing-numbers-through-10-example
mp4 & srt file name: **tJrSILRXOUc**

**Comment by jamalex** [comment]
Interesting... I couldn't get a video page with (English) subtitles available to render at all: #3861

**Comment by radinamatic** [comment]
This must be a URL issue, because if I copy the srt file in the "content" folder and hardcode this:

`<track kind="captions" src="/content/tJrSILRXOUc.srt" srclang="en" label="English" />`

inside the video element in the \hbtemplates\video\video-player.handlebars file, I can see the captions just fine...

**Comment by rtibbles** [comment]
It may well be a similar cause to #3861 then - hopefully @jamalex's fix will solve it.

**Comment by radinamatic** [comment]
I tried that as soon as I saw his issue, but here  `<track kind="captions" src="{{ this.url }}" ` did not change a thing...

**Comment by rtibbles** [comment]
Yours must be an issue with the URL itself not being properly put onto the video content.

Are you using the DO_NOT_RELOAD_CONTENT_CACHE_AT_STARTUP flag?

If so, it's possible this is not being properly invalidated by subtitle download.

**Comment by radinamatic** [comment]
Nope, I have nothing in the local settings apart from `DEBUG = False`

**Comment by radinamatic** [comment]
Can somebody please try to replicate this? I pulled today's develop, after @jamalex merged his #3864, and as suspected, I still can't see the captions button... :(

**Comment by MCGallaspy** [comment]
Yes, I can indeed confirm that subtitles are not displaying on `develop`. I have a blank `local_settings.py`, I downloaded the English subtitles (so theoretically every video should have them) and I see no widget on the video player.

**Comment by radinamatic** [comment]
Phew, feels good not to be the alone... ;)

**Comment by MCGallaspy** [comment]
Okay, after looking in to this for a while:
* Maybe I just missed it the first time around (the CC button is really small!), but I do have English subtitles on videos when DEBUG=True. (Not sure if DEBUG makes a difference here)
* I did notice that running with the django debug server did not serve the srt files (`kalite manage kaserve`), but that using cherrypy did (`kalite manage kaserve --production`). @radinamatic are you not seeing the option for subtitles on videos where you know the subtitle does exist? Does it make a difference if you use the django debug server vs cherrypy?

**Comment by radinamatic** [comment]
So far I did not use neither django nor cherrypy debug server (unless they run without my knowledge), are they necessary to view videos? 
I did try with several videos for which I for sure had subtitles downloaded, and the only time I saw the CC button is (see above) when I hardcoded the link for a specific srt file...

CC button is the same size as the video speed button, so that might be something that we might consider changing.

**Comment by MCGallaspy** [comment]
@radinamatic you must have been using one or the other :)

How did you start the server? Basically, that command will determine which webserver is used (cherrypy or django debug server).

**Comment by radinamatic** [comment]
`bin/kalite start` or `bin\windows\kalite.bat start`
So "debug" server != "normal" server ( I'm feeling so noob right now :(

**Comment by rtibbles** [comment]
@MCGallaspy The DEBUG=True issue is because the srt files are saved directly into the static directory and hence when you run with DEBUG=True, and it serves files directly from the INSTALLED_APPS static dirs, the srt files are not there (similar to why the compiled django_js_reverse JS file does not work in development mode).

**Comment by rtibbles** [comment]
@radinamatic Generally, if you run with DEBUG=TRUE static files are being served by Django. If you run with DEBUG=FALSE, then static files are being served by CherryPy.

**Comment by MCGallaspy** [comment]
@radinamatic @rtibbles
>Generally, if you run with DEBUG=TRUE static files are being served by Django. If you run with DEBUG=FALSE, then static files are being served by CherryPy.

Right, except that if you use `bin/kalite start` then you will be serving with cherrypy regardless of your DEBUG setting.

**Comment by radinamatic** [comment]
Ok, I'm learning a lot tight now... So what's the bottom line? What should I use, because I'm still not seeing srt files in any way, shape or form...

I'm running `kalite manage kaserve` right now and it's giving a lot of cherrypy errors...

**Comment by MCGallaspy** [comment]
@radinamatic I think the best policy is to run the server using either `kalite start` or `kalite manage kaserve --production`. Both should use cherrypy; the former will run in the background, and the latter will print to stdout/stderr.

**Comment by radinamatic** [comment]
Ok, just tried with `kalite start` only, no subs, let me try `kalite manage kaserve --production`.

**Comment by MCGallaspy** [comment]
@radinamatic I wouldn't expect different results with those commands, just different output to the terminal

**Comment by MCGallaspy** [comment]
@rtibbles 
At what point in the language pack download process are subtitles unpacked? I'm looking for srt files corresponding to the content item `/math/early-math/cc-early-math-counting-topic/cc-early-math-comparing-numbers/v/comparing-numbers-through-10-example` and I'm getting these results:

```sh
>> find . -name "__nkbr6DeTg.srt"
./kalite/static/srt/en/subtitles/__nkbr6DeTg.srt
./locale/es/subtitles/__nkbr6DeTg.srt
```

English subtitles work fine on the video, español does not. I suspect @radinamatic has this issue because subtitles are not in the expected place.

**Comment by rtibbles** [comment]
Ah.

**Comment by rtibbles** [comment]
@MCGallaspy line 188 move_srts function.

**Comment by rtibbles** [comment]
languagepackdownload.py in updates management commands.

**Comment by radinamatic** [comment]
@MCGallaspy ATM I have English configured as default lang, so video should load English srt. 
On my (Windows) machine, srt files are in:

ka-lite\kalite\static\srt\en\subtitles &
ka-lite\kalite\static\srt\es\subtitles

**Comment by MCGallaspy** [comment]
@radinamatic if the srt files do indeed exist and are being served (can you get to them if you enter the address in the browser? or does it return a 404 response?), then I'm at a bit of a loss as to why you're not able to select them.

**Comment by rtibbles** [comment]
@radinamatic had previously reported they work if the tag was added manually, which is why I thought it was a stamping issue.

**Comment by MCGallaspy** [comment]
The annotation seems to work for me (I inspected the code pretty closely), but my language pack download seems to have failed somehow, too. (My srts are not in the right place.)

**Comment by radinamatic** [comment]
OTOH, running `kalite manage kaserve --production` got me some cherrypy errors, but still no subtitles on either FF or Chrome:
```
$ kalite manage kaserve --production
d:\github15\KA\ka-lite\kalite\settings\base.py:38: RemovedInKALite_v015_Warning:
 We will be deprecating the old way of statically importing custom settings, in
favor of a more flexible way. The easiest way to update your installation is to
rename your local_settings.py (keeping it in the same directory) and add an impo
rt statement in the very first line of the new file so it looks like this:

    from kalite.settings.base import *
    # Put custom settings here...
    FOO = BAR

and then call kalite start with an additional argument pointing to your new sett
ings module:

    kalite start --settings=kalite.my_settings

In the future, it is recommended not to keep your own settings module in the kal
ite code base but to put the file somewhere else in your python path, for instan
ce in the current directory when running 'kalite --settings=my_module'.
  RemovedInKALite_v015_Warning
INFO:kalite:Copying static media...
WARNING:py.warnings:d:\github15\KA\ka-lite\python-packages\tastypie\resources.py
:328: UserWarning: 'override_urls' is a deprecated method & will be removed by v
1.0.0. Please rename your method to ``prepend_urls``.
  warnings.warn("'override_urls' is a deprecated method & will be removed by v1.
0.0. Please rename your method to ``prepend_urls``.")

To access KA Lite from another connected computer, try the following address(es)
:
        http://192.168.1.33:8008/
        http://192.168.56.1:8008/
To access KA Lite from this machine, try the following address:
        http://127.0.0.1:8008/
[17/Jun/2015:23:39:04] ENGINE Listening for SIGTERM.
INFO:cherrypy.error:[17/Jun/2015:23:39:04] ENGINE Listening for SIGTERM.
[17/Jun/2015:23:39:04] ENGINE Bus STARTING
INFO:cherrypy.error:[17/Jun/2015:23:39:04] ENGINE Bus STARTING
[17/Jun/2015:23:39:04]  Loading and serving the Django application
INFO:cherrypy.error:[17/Jun/2015:23:39:04]  Loading and serving the Django appli
cation
[17/Jun/2015:23:39:04] ENGINE Started monitor thread '_TimeoutMonitor'.
INFO:cherrypy.error:[17/Jun/2015:23:39:04] ENGINE Started monitor thread '_Timeo
utMonitor'.
[17/Jun/2015:23:39:04] ENGINE Serving on 0.0.0.0:8008
INFO:cherrypy.error:[17/Jun/2015:23:39:04] ENGINE Serving on 0.0.0.0:8008
[17/Jun/2015:23:39:04] ENGINE Bus STARTED
INFO:cherrypy.error:[17/Jun/2015:23:39:04] ENGINE Bus STARTED
192.168.1.33 - - [17/Jun/2015:23:41:06] "GET /static/django_js_reverse/js/revers
e.js HTTP/1.1" 200 130141 "http://192.168.1.33:8008/" "Mozilla/5.0 (Windows NT 5
.1; rv:38.0) Gecko/20100101 Firefox/38.0"
INFO:cherrypy.access.36600592:192.168.1.33 - - [17/Jun/2015:23:41:06] "GET /stat
ic/django_js_reverse/js/reverse.js HTTP/1.1" 200 130141 "http://192.168.1.33:800
8/" "Mozilla/5.0 (Windows NT 5.1; rv:38.0) Gecko/20100101 Firefox/38.0"
192.168.1.33 - - [17/Jun/2015:23:41:10] "GET /static/perseus/ke/third_party/Math
Jax/2.1/MathJax.js?config=/static/perseus/ke/third_party/MathJax/2.1/config/KAth
Jax-8f02f65cba7722b3e529bd9dfa6ac25d HTTP/1.1" 304 - "http://192.168.1.33:8008/l
earn/" "Mozilla/5.0 (Windows NT 5.1; rv:38.0) Gecko/20100101 Firefox/38.0"
INFO:cherrypy.access.36600592:192.168.1.33 - - [17/Jun/2015:23:41:10] "GET /stat
ic/perseus/ke/third_party/MathJax/2.1/MathJax.js?config=/static/perseus/ke/third
_party/MathJax/2.1/config/KAthJax-8f02f65cba7722b3e529bd9dfa6ac25d HTTP/1.1" 304
 - "http://192.168.1.33:8008/learn/" "Mozilla/5.0 (Windows NT 5.1; rv:38.0) Geck
o/20100101 Firefox/38.0"
```

**Comment by MCGallaspy** [comment]
Those aren't errors, that's the logging output of the server.

**Comment by rtibbles** [comment]
No errors there, just deprecation warnings.

**Comment by radinamatic** [comment]
Sorry, got blindsided by "INFO:cherrypy.error:" in there...

**Comment by MCGallaspy** [comment]
So I can't reproduce this error, and the only other thing I can think is to observe @radinamatic experiencing this issue in order to figure out what the "missing link" is.

**Comment by rtibbles** [comment]
Stepping through the annotation process seems the most likely to bear fruit.

**Comment by radinamatic** [comment]
Which path should I use to test if srt can be opened directly in the browser? If I copy them in the content folder with .mp4 files, I can see them in the browser...

**Comment by MCGallaspy** [comment]
We discovered this is an os-specific issue due to how pathnames are handled in `topic_tools/__init__.py`

**Comment by rtibbles** [comment]
This should now be fixed on develop, fixed by #3897.

**Comment by radinamatic** [comment]
Ok, have been testing this for a while now and I can confirm that the issue can be considered fixed. There are a few things I'd like to mention:

- I could see all the captions in English, but I also stumbled on one that did not display despite the fact srt file was present in my srt\en\subtitles folder. I couldn't figure out why until I actually opened the file itself and saw that the timecodes were all wrong - essentially all the subtitle lines had the `99:59:59,999 --> 99:59:59,999` timecode, so no wonder they did not show up. The caption in question was T0zpF_j7Mvo.srt for [this](https://www.khanacademy.org/science/physics/one-dimensional-motion/kinematic_formulas/v/plotting-projectile-displacement-acceleration-and-velocity) video. Correcting this falls outside our present purview, but I'd like to know where/how to report this to be corrected.

- On certain videos Spanish sub is reported as available in the  [APRIL 2015 MASTER](https://docs.google.com/spreadsheets/d/1k5xh2UXV3EchRHnYzeP6-YGKrmba22vsltGSuU9bL88/edit#gid=0) file, but would not show. Subtitle files in question were not present in the **srt\es\subtitles folder** so there is no error in the code, but likely in the Google spreadsheet. This is also not an urgent matter, and I do not know if/how the spreadsheet itself is related to the subs update packages in the Admin's Manage tab. However, if it is, we should keep it in mid as we try to (re)implement the Subtitle availability indication #3889 as per @MCGallaspy proposal. 

- [APRIL 2015 MASTER](https://docs.google.com/spreadsheets/d/1k5xh2UXV3EchRHnYzeP6-YGKrmba22vsltGSuU9bL88/edit#gid=0) file itself is enormous and extremely difficult to manage as it makes my browser unresponsive. I had to make a copy and delete the columns that I didn't need in order to be able to work with it. Just wondering if anybody has an idea how to deal with this...
